tools/shared.py

```diff
@@ -1627,20 +1584,85 @@ def is_ar(filename):
 
   @staticmethod
   def is_bitcode(filename):
-    # look for magic signature
-    b = open(filename, 'r').read(4)
-    if b[0] == 'B' and b[1] == 'C':
-      return True
-    # look for ar signature
-    elif Building.is_ar(filename):
-      return True
-    # on OS X, there is a 20-byte prefix
-    elif ord(b[0]) == 222 and ord(b[1]) == 192 and ord(b[2]) == 23 and ord(b[3]) == 11:
-      b = open(filename, 'r').read(24)
-      return b[20] == 'B' and b[21] == 'C'
+    try:
+      sigcheck = False
+      # look for magic signature
+      with open(filename, 'r') as f:
+        b = f.read(4)
+        if len(b) < 4: return False
+        if len(b) >= 2 and b[0] == 'B' and b[1] == 'C':
+          sigcheck = True
+        # look for ar signature
+        elif Building.is_ar(filename):
+          sigcheck = True
+        # on OS X, there is a 20-byte prefix
+        elif len(b) == 4 and ord(b[0]) == 222 and ord(b[1]) == 192 and \
+                             ord(b[2]) == 23 and ord(b[3]) == 11:
+          b = f.read(18)
+          sigcheck = len(b) == 18 and b[16] == 'B' and b[17] == 'C'
+      return sigcheck
+    except Exception, e:
+      logging.debug('Building.is_bitcode failed on file \'%s\'! Failed on exception: %s' % (filename, e))
+      return False
 
+  @staticmethod
+  def is_jslibrary(filename):
+    if not filename.endswith(('.jso','.jso.o')):
+      return False
+    try:
+      with open(filename, 'r') as f:
```

Comment:
this part is very worrying to me. why do we need this hack?

i would think we should remove this and just check the suffixes in `emcc`, the only place this is used. was there a reason not to do that?